kgforge/core/archetypes/database.py

```diff
@@ -0,0 +1,146 @@
+# 
+# Blue Brain Nexus Forge is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+# 
+# Blue Brain Nexus Forge is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser
+# General Public License for more details.
+# 
+# You should have received a copy of the GNU Lesser General Public License
+# along with Blue Brain Nexus Forge. If not, see <https://choosealicense.com/licenses/lgpl-3.0/>.
+
+from abc import ABC, abstractmethod
+import json
+from pathlib import Path
+from typing import Any, Optional, Callable, Dict, List, Union
+
+from kgforge.core import Resource
+from kgforge.core.commons.context import Context
+from kgforge.core.commons.execution import not_supported
+from kgforge.core.archetypes import Mapping, Model
+from kgforge.core.commons.attributes import repr_class
+from kgforge.core.commons.exceptions import ConfigurationError
+from kgforge.core.commons.dictionaries import use_values
+from kgforge.core.commons.imports import import_class
+
+
+class Database(ABC):
+
+    # POLICY Methods of archetypes, except __init__, should not have optional arguments.
+
+    # POLICY Implementations should be declared in kgforge/specializations/databases/__init__.py.
+    # POLICY Implementations should not add methods but private functions in the file.
+    # POLICY Implementations should pass tests/specializations/databases/test_databases.py.
+
+    def __init__(self, forge : Optional["KnowledgeGraphForge"], source: str, **config) -> None:
+        # POLICY Resolver data access should be lazy, unless it takes less than a second.
+        # POLICY There could be data caching but it should be aware of changes made in the source.
+        self._forge: Optional["KnowledgeGraphForge"] = forge
+        # Model
+        model_config = config.pop("model")
+        if model_config["origin"] == "store":
+            use_values(
+                model_config,
+                config,
+                ["endpoint", "token", "bucket", "vocabulary"],
+            )
+        model_name = model_config.pop("name")
+        model = import_class(model_name, "models")
+        self._model: Model = model(**model_config)
+        self.context = self._model.context()
+        if 'model_context' not in config:
+            config['model_context'] = self._model.context()
+        self.source: str = source
+        self.service: Any = self._initialize_service(self.source, **config)
+
+    def __repr__(self) -> str:
+        return repr_class(self)
+
+    def map(self, resources : Union[List[Union[Resource, str]], Union[Resource, str]],
```

Comment:
- If I understand correctly, if type_ is not provided, a resource is mapped according to the mapping that is associated with its (the resource’s) type property by model. 
- The if else on the resource type assumes that if it’s not of type Resource, it’s of type Dict, but the typing in map defines it to be a Union (or list of) of Resource and str. Should it be dict instead? 
- If a resource has no type_ and no type is provided, there is a silent error? 
- Similarly there is a series of ifs related to type_ and if neither of the cases that are “valid” are true then it’s silently considering the resource as mapped. Shouldn’t there be some warning? Similar to with the update/register operations, it gives the count of successful operations and failed ones and error messages linked to the failures. 
- (In the last two points I’ve talked about “errors” but really it’s just considering a resource to be mapped when it’s not been. However, the thing that would happen is at least that if the element provided was a dictionary it would be converted into a Resource)